pgrx/src/fn_call.rs

```diff
@@ -336,69 +337,79 @@ fn lookup_fn(fname: &str, args: &[&dyn FnCallArg]) -> Result<pg_sys::Oid> {
     // function following the normal SEARCH_PATH rules, ensuring its argument type Oids
     // exactly match the ones from the user's input arguments.  It does not evaluate the
     // return type, so we'll have to do that later
-    let mut parts_list = List::<*mut std::ffi::c_void>::default();
-    let result = PgTryBuilder::new(AssertUnwindSafe(|| unsafe {
-        let arg_types = args.iter().map(|a| a.type_oid()).collect::<Vec<_>>();
-        let nargs: i16 = arg_types.len().try_into().map_err(|_| FnCallError::TooManyArguments)?;
-
-        // parse the function name into its possibly-qualified name parts
-        let ident_parts = parse_sql_ident(&fname)?;
-        ident_parts
-            .iter_deny_null()
-            .map(|part| {
-                // SAFETY:  `.as_pg_cstr()` palloc's a char* and `makeString` just takes ownership of it
-                pg_sys::makeString(part.as_pg_cstr())
-            })
-            .for_each(|part| {
-                parts_list.unstable_push_in_context(part.cast(), pg_sys::CurrentMemoryContext);
-            });
-
-        // look up an exact match based on the exact number of arguments we have
-        //
-        // SAFETY:  we've allocated a PgList with the proper String node elements representing its name
-        // and we've allocated Vec of argument type oids which can be represented as a pointer.
-        let mut fnoid =
-            pg_sys::LookupFuncName(parts_list.as_mut_ptr(), nargs as _, arg_types.as_ptr(), true);
-
-        if fnoid == pg_sys::InvalidOid {
-            // if that didn't find a function, maybe we've got some defaults in there, so do a lookup
-            // where Postgres will consider that
-            fnoid = pg_sys::LookupFuncName(
+    mem::current_context(|mcx| {
```

Comment:
The need to do this, and the problems with memory leaks in #1341, strongly suggests we should be accepting a MemCx argument to this function and also the public interface of `fn_call`.